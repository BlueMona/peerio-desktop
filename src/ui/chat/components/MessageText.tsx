import React from 'react';
import { observer } from 'mobx-react';
import { Message } from 'peerio-icebear/dist/models';
import { chatSchema, Renderer } from '~/helpers/chat/prosemirror/chat-schema';
import { User } from 'peerio-icebear';

let legacyProcessMessageForDisplay: (msg: Message) => { __html: string };

interface MessageTextProps {
    /** icebear message */
    message: Message;

    /** handler of event when user clicks on contact */
    onClickContact: (ev: React.MouseEvent<Element>) => void;
}

@observer
export default class MessageText extends React.Component<MessageTextProps> {
    render() {
        const { message, onClickContact } = this.props;

        if (!message) return null;

        const richText = message.richText;
        if (
            richText &&
            typeof richText === 'object' &&
            // @ts-ignore needs icebear fixes
            richText.type === 'doc' &&
            // @ts-ignore needs icebear fixes
            richText.content
        ) {
            try {
                // Creating the ProseMirror node from the JSON may seem like an
                // added step/layer of indirection, but it lets us validate the
                // rich text payload and ensure it conforms to the schema.
                const proseMirrorNode = chatSchema.nodeFromJSON(richText);

                // Note that an error in the renderer component won't get caught
                // by this try-catch -- it's not actually invoked in this stack
                // frame.
                return (
                    <Renderer
                        fragment={proseMirrorNode.content}
                        onClickContact={onClickContact}
                        currentUser={User.current.username}
                    />
                );
            } catch (e) {
                console.warn(`Couldn't deserialize message rich text:`, e);
            }
        }

        if (typeof message.text !== 'string') {
            // HACK: React error boundaries only catch errors in children, so we
            // wrap this throw in a createElement.
            return React.createElement(() => {
                throw new Error("Can't render rich text and message has no plaintext!");
            });
        }

        // lazily load legacy message processing logic. the emojione file it
        // requires is pretty huge and slows startup, so we should only require
        // it if it's necessary.
        legacyProcessMessageForDisplay =
            legacyProcessMessageForDisplay ||
            require('~/helpers/chat/legacy-process-message-for-display').processMessageForDisplay;
        return (
            <p
                // eslint-disable-next-line react/no-danger
                dangerouslySetInnerHTML={legacyProcessMessageForDisplay(message)}
                className="selectable"
            />
        );
    }
}
