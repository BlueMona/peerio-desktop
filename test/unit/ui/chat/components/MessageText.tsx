import React from 'react';
import { shallow, mount, render } from 'enzyme';

import MessageText from '~/ui/chat/components/MessageText';

const onClickContact = () => {};

function makeRichTextMessage(content: any[]): any {
    return {
        richText: {
            type: 'doc',
            content: [
                {
                    type: 'paragraph',
                    content
                }
            ]
        }
    };
}

describe('<MessageText>', () => {
    it('should match the snapshot for all fixtures', () => {
        ['1', '2', '3', '4', '5']
            .map(fn => require(`./message-text-fixtures/${fn}.json`))
            .forEach(richText => {
                expect(
                    render(
                        <MessageText
                            message={{ richText } as any}
                            onClickContact={onClickContact}
                            currentUsername={'testuser'}
                        />
                    )
                ).toMatchSnapshot();
            });
    });

    it('should render nothing if passed an empty message', () => {
        expect(
            shallow(
                <MessageText
                    message={null}
                    onClickContact={onClickContact}
                    currentUsername={'testuser'}
                />
            )
        ).toBeEmptyRender();
    });

    it('should contain a highlight if the message contains a mention of the current user', () => {
        const username = 'testuser';

        const message = makeRichTextMessage([
            {
                type: 'mention',
                attrs: {
                    username
                }
            }
        ]);

        expect(
            mount(
                <MessageText
                    message={message}
                    onClickContact={onClickContact}
                    currentUsername={username}
                />
            )
        ).toContainMatchingElement('.self');
    });

    it('should render jumboji if the message contains only {1,3} emoji', () => {
        const emojiNode = {
            type: 'emoji',
            attrs: {
                shortname: ':apple:'
            }
        };

        [[emojiNode], [emojiNode, emojiNode], [emojiNode, emojiNode, emojiNode]]
            .map(makeRichTextMessage)
            .forEach(message => {
                expect(
                    mount(
                        <MessageText
                            message={message}
                            onClickContact={onClickContact}
                            currentUsername={'testUser'}
                        />
                    )
                ).toContainMatchingElement('.jumboji');
            });
    });

    it('should not render jumboji if the message contains text or more than three emoji', () => {
        const emojiNode = { type: 'emoji', attrs: { shortname: ':apple:' } };

        [
            [{ type: 'text', text: 'just text' }],
            [{ type: 'text', text: 'text and emoji ' }, emojiNode],
            [emojiNode, emojiNode, emojiNode, emojiNode]
        ]
            .map(makeRichTextMessage)
            .forEach(message => {
                expect(
                    mount(
                        <MessageText
                            message={message}
                            onClickContact={onClickContact}
                            currentUsername={'testUser'}
                        />
                    )
                ).not.toContainMatchingElement('.jumboji');
            });
    });

    it('should process legacy messages if no rich text is provided', () => {
        jest.mock('~/helpers/chat/legacy-process-message-for-display', () => ({
            processMessageForDisplay: jest.fn(msg => ({ __html: `<div>${msg.text}</div>` }))
        }));

        const message = {
            text: 'a legacy message.'
        };

        expect(
            render(
                <MessageText
                    message={message as any}
                    onClickContact={onClickContact}
                    currentUsername={'testUser'}
                />
            ).text()
        ).toContain(message.text);

        expect(
            require('~/helpers/chat/legacy-process-message-for-display').processMessageForDisplay
        ).toHaveBeenCalledTimes(1);
    });
});
