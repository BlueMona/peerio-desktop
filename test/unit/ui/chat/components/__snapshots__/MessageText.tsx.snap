// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageText> should match the snapshot for all fixtures 1`] = `
<div
  class=""
>
  <p>
    rich 
    <strong>
      rich 
    </strong>
    <em>
      <strong>
        very rich 
      </strong>
    </em>
    <s>
      test
    </s>
     text
  </p>
</div>
`;

exports[`<MessageText> should match the snapshot for all fixtures 2`] = `
<div
  class=""
>
  <p>
    yo 
    <img
      alt="😎"
      class="emojione"
      src="./static/emoji/png/1f60e.png"
      title=":sunglasses:"
    />
  </p>
</div>
`;

exports[`<MessageText> should match the snapshot for all fixtures 3`] = `
<div
  class=""
>
  <p>
    <img
      alt="🍎"
      class="emojione"
      src="./static/emoji/png/1f34e.png"
      title=":apple:"
    />
    <img
      alt="🍏"
      class="emojione"
      src="./static/emoji/png/1f34f.png"
      title=":green_apple:"
    />
    <img
      alt="🍈"
      class="emojione"
      src="./static/emoji/png/1f348.png"
      title=":melon:"
    />
    <img
      alt="🥥"
      class="emojione"
      src="./static/emoji/png/1f965.png"
      title=":coconut:"
    />
  </p>
</div>
`;

exports[`<MessageText> should match the snapshot for all fixtures 4`] = `
<div
  class="jumboji"
>
  <p>
    <img
      alt="😚"
      class="emojione"
      src="./static/emoji/png/1f61a.png"
      title=":kissing_closed_eyes:"
    />
    <img
      alt="🏈"
      class="emojione"
      src="./static/emoji/png/1f3c8.png"
      title=":football:"
    />
  </p>
</div>
`;

exports[`<MessageText> should match the snapshot for all fixtures 5`] = `
<div
  class=""
>
  <p>
    hey 
    <span
      class="mention clickable"
      data-username="butt"
    >
      @butt
    </span>
     how's... 
    <em>
      life
    </em>
  </p>
</div>
`;
